source/distributed/fully_distributed_tria_util.cc

```diff
@@ -0,0 +1,510 @@
+                }
+
+            // 3) enumerate locally relevant vertices
+            unsigned int vertex_counter = 0;
+            for (auto &vertex : vertices_locally_relevant)
+              {
+                vertices.push_back(tria.get_vertices()[vertex.first]);
+                vertex.second = vertex_counter++;
+              }
+
+            // 4) correct vertices of cells (make them local)
+            for (auto &cell : cells)
+              for (unsigned int v = 0; v < GeometryInfo<dim>::vertices_per_cell;
+                   v++)
+                cell.vertices[v] = vertices_locally_relevant[cell.vertices[v]];
+          }
+        else
+          {
+            // 1) collect locally relevant cells (set user_flag)
+
+            // 1a) clear user_flags
+            for (auto &cell : tria)
+              cell.recursively_clear_user_flag();
+
+            // 1b) loop over levels (from fine to coarse) and mark on each level
+            //     the locally relevant cells
+            for (unsigned int level =
+                   tria.get_triangulation().n_global_levels() - 1;
+                 level != numbers::invalid_unsigned_int;
+                 level--)
+              {
+                // collect vertices connected to a (on any level) locally owned
+                // cell
+                std::set<unsigned int> vertices_owned_by_loclly_owned_cells;
+                for (auto cell : tria.cell_iterators_on_level(level))
+                  if (cell->level_subdomain_id() == my_rank ||
+                      (cell->active() && cell->subdomain_id() == my_rank))
+                    add_vertices_of_cell_to_vertices_owned_by_loclly_owned_cells(
+                      cell, vertices_owned_by_loclly_owned_cells);
+
+                // helper function to determine if cell is locally relevant
+                // (i.e. a cell which is connected via a vertex via a locally
+                // owned cell)
+                auto is_locally_relevant = [&](auto &cell) {
+                  for (unsigned int v = 0;
+                       v < GeometryInfo<dim>::vertices_per_cell;
+                       v++)
+                    if (vertices_owned_by_loclly_owned_cells.find(
+                          cell->vertex_index(v)) !=
+                        vertices_owned_by_loclly_owned_cells.end())
+                      return true;
+                  return false;
+                };
+
+                // mark all locally relevant cells
+                for (auto cell : tria.cell_iterators_on_level(level))
+                  if (is_locally_relevant(cell))
+                    set_user_flag_reverse(cell);
+              }
+
+            // 2) setup coarse-grid triangulation
+            {
+              std::map<unsigned int, unsigned int> vertices_locally_relevant;
+
+              // a) loop over all cells
+              for (auto cell : tria.cell_iterators_on_level(0))
+                {
+                  if (!cell->user_flag_set())
```

Comment:
Since you're using the user flags, I think you should save/restore the user flags. There are many places in the library where we do this to ensure that users don't see what we're doing with these flags.